CLI: Modularize katello-agent and errata tests - issue #2199

abalakh opened the pull request:
Modularized #1708 tests for simple re-use in other katello-agent- or errata- related tests.
The task idea is fully described [here (comment to PR #2186)](https://github.com/SatelliteQE/robottelo/pull/2186#issuecomment-100028625).
Test results:
```
nosetests tests/foreman/cli/test_contenthost.py -m test_contenthost_
..
----------------------------------------------------------------------
Ran 2 tests in 721.204s

OK
```
Resolves #2199

abalakh commented:
Placed [DO NOT MERGE] tag as I don't think factory is the best place for ``activationkey_subscribe_to_custom_repo()``` and ```activationkey_subscribe_to_rh_repo()```, but it's available outside the content host tests (can be used in errata tests), have almost all dependencies imported (so i'll not overload some another commonly used lib with a lot of imports) and can raise exceptions in case something went wrong. Please let me know if there's a better place to put those functions in.

abalakh commented:
I also tried to split those functions differently, to some ```def make_activationkey_with_repo_subscription()``` to fit the factory code style, but wasn't successful, as in tests we need to add subscription for repo at least twice, within the same activation-key, CV, env and organization, so functions should update existing objects like act key, CV, env, and not create the new ones.

oshtaier commented:
yeah, it seems that we need an advice from someone experienced :). Mentioned ```activationkey_subscribe_to_custom_repo()``` and ```activationkey_subscribe_to_rh_repo()``` methods are definitely out of style and ideology of the code which is surrounded it

sthirugn commented on robottelo/vm.py:
```diff
@@ -171,6 +171,80 @@ def run(self, cmd):
+            raise VirtualMachineError(
+                'Failed to install {0} rpm.'.format(package_name)
+            )
+
```
I like all the functions defined in vm.py :+1:

sthirugn commented on robottelo/cli/factory.py:
```diff
@@ -109,6 +114,226 @@ def create_object(cli_object, options, values):
 
 
 @cacheable
+def activationkey_subscribe_to_custom_repo(options=None):
```
the name is misleading and is not conveying what this function is trying to do - May be we can call this something like:
```
setup_org_for_a_custom_repo(options=None)
```

Also the function's docstring should explain what this function is trying to do, something like:
```
"""
Sets up Org for the given custom repo by:
1. Checks for an organization-id else it creates a new org
2. Creates a new product with the custom repo.  Synchronizes the repo.
3. Checks for a content-view-id or creates a new Content view and 
    - adds the custom repo
    - publishes
    - promotes  to the activation key's lifecycle environment
4. Adds the custom repo subscription to the Activation key

sthirugn commented on robottelo/cli/factory.py:
```diff
@@ -109,6 +114,226 @@ def create_object(cli_object, options, values):
+
+@cacheable
+def activationkey_subscribe_to_rh_repo(options=None):
+    """
```
Same comment as the other function above.

sthirugn commented:
Regarding where to keep `activationkey_subscribe_to_custom_repo()` and `activationkey_subscribe_to_rh_repo()` IMO these can stay in factory.py or perhaps move to org.py but that is debatable.  Probably lets clean the code first to make it clean and we can decide on a place.

elyezer commented:
I think the factory is the good place for these kind of functions and can serve as a base place. I think the version 2 of the CLI factories would be a way to create a object and make it create the dependents too, like nailgun already does.

Also if we need something to be used by API/CLI/UI we can definitely be placed on common/helpers.py but I vote for creating them using API calls because they are faster. Also the objective of these helpers are just setup entities in order to get an environment to check other things.

abalakh commented:
OK, thanks for clarification. So i've updated factory functions names, docstrings, re-arranged vm functions positions, made a small fix there and rebased everything.
Think now PR is ready for review, removing '[DO NOT MERGE]' tag :)

sthirugn commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,242 @@ def make_template(options=None):
+
+    Args::
+
+        activationkey-id - ID of activation key
```
:+1: I like the changes, but can we also make activation key optional?

elyezer commented:
ACK, pending @sthirugn suggestion.

abalakh commented:
Updated. Made ```activationkey-id``` optional, added optional arg ```lifecycleenvironment-id```, updated rpm querying according to #2262, fixed small issues, updated docstrings and made rebase.
Test results:
```
nosetests tests/foreman/cli/test_contenthost.py -m test_contenthost_
..
----------------------------------------------------------------------
Ran 2 tests in 788.971s

OK
```
@sthirugn, you may want to review it once more, since there're some changes in ```setup_org_for_a_custom_repo()``` and ```setup_org_for_a_rh_repo()``` logic. Thanks :)

elyezer commented:
ACK. Will leave for @sthirugn take another look and merge it.

sthirugn commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,279 @@ def make_template(options=None):
+            if result.return_code != 0:
+                raise CLIFactoryError(
+                    'Failed to add subscription to activation key'
+                )
```
Subscription part is redundant between setup_org_for_a_custom_repo and setup_org_for_a_rh_repo.  This can be a separate function which both the other functions can call.  But this can be done in next PR.

elyezer replied:
:+1:

abalakh replied:
So you are suggesting some ```activationkey_add_subscription_to_repo()``` function, right?
Will this function be used from somwhere outside? I mean, I'm OK with that change if some tests will use this function, but if won't - I'm not sure if it's good approach to split the code to functions, which are used just once.

sthirugn replied:
I am pretty sure that there will be more use cases coming up for adding the given subscription(s) to an activationkey.

abalakh replied:
OK, great, will work on that. No need in new PRs :)

sthirugn commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -567,115 +559,38 @@ def setUp(self):
 
```
Do we need to run the setUp()  separately before test_contenthost_get_errata_info() and test_contenthost_apply_errata()? Is it not enough to run setUp() once in this class? I am trying to save time from running redundant setup steps.

abalakh replied:
I've done that on purpose - currently we have 2 tests, which are trying to install some package manually, but future content host tests, for example, tests from issues #1709 and #1710, will verify automatic package installation/upgrade/deletion. That means that I need VM without some specific package installed for each of these tests to avoid conflicts.

sthirugn replied:
If I understand it right, we need to move the vm code from setup to individual tests to handle:
```py
self.vm = VirtualMachine(distro='rhel71')
self.vm.create()
self.vm.install_katello_cert()
self.vm.register_contenthost(
self.activation_key['name'], self.org['label']
)
# Install katello-agent
self.vm.enable_repo(REPOS['rhst7']['id'])
self.vm.install_katello_agent()
```
If we move this code to the tests, can we reuse setup?  By this way, tests can create their own vms, but use all the other pre-requisites from existing org, ak, content-view, publish, promote, etc?

abalakh replied:
If code is used in the beginning of each test of the class, than it _should_ be moved to setUp(), that what setUp is made for :)
What are you trying to suggest is probably not ```setUp()```, but ```setUpClass()```, isn't it?
In this case, code can be splitted to something like that:
```python
class TestContentHostKatelloAgent(CLITestCase):

    def setUpClass():
    # Create org, env, CV, AK

    def setUp():
    # Create VM, install katello-agent

    def tearDown():
    # Destroy VM

    def test_1():
    # Testing here
```
This will allow us to use 1 org, env and AK for all the tests and we won't duplicate VM code inside each test. Did I understand you correctly?

abalakh replied:
If that's the idea, then we have benefit in saving time by using the same org/env/CV, but the drawbacks are increased code complexity and changing the code style. I've quickly searched through code and couldn't found classes, which are using both ```setUpClass()``` and ```setUp()```, so it's definitely not the commonly-used approach. @elyezer , could you please share your thoughts on that?

sthirugn replied:
Yes, Sorry I didn't mention setUpClass explicity!  But I thought of just using setUpClass and move vm creation part to the tests or perhaps as an other small function which the tests can call.  But your approach of using setUpClass and setUp looks nice as well.

@elyezer What do you think about this?

elyezer replied:
@abalakh and @sthirugn I don't see any issue on using the setUpClass and setUp methods on the same TestCase class if they are used the way they should be used.

There is just one concern here, if something fail on setUpClass then the entire class will not run while if setUp fail just that specific test will not run.

An approach to workaround this would be creating class attributes, for example:

```python
class MyTestCase(UITestCase):
    org = None
    env = None
    content_view = None
    activation_key = None

    def setUp(self):
        if MyTestCase.org is None:
             MyTestCase.org = create_org() # if fail then MyTestCase.org should be set to None
        # ...
```

elyezer replied:
Actually the behaviour will depend on the test runner, on py.test:

```
py.test test.py
=================================== test session starts ===================================
platform darwin -- Python 2.7.6 -- py-1.4.26 -- pytest-2.6.4
collected 2 items

test.py EE

========================================= ERRORS ==========================================
_______________________________ ERROR at setup of A.test_1 ________________________________

cls = <class 'test.A'>

    @classmethod
    def setUpClass(cls):
>       raise Exception('Setup class failed')
E       Exception: Setup class failed

test.py:6: Exception
_______________________________ ERROR at setup of A.test_2 ________________________________

cls = <class 'test.A'>

    @classmethod
    def setUpClass(cls):
>       raise Exception('Setup class failed')
E       Exception: Setup class failed

test.py:6: Exception
================================= 2 error in 0.03 seconds =================================
```

On nosetests:

```
nosetests test.py
E
======================================================================
ERROR: test suite for <class 'test.A'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/elyezer/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/Users/elyezer/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/Users/elyezer/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/Users/elyezer/.virtualenvs/robottelo/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/Users/elyezer/code/robottelo/test.py", line 6, in setUpClass
    raise Exception('Setup class failed')
Exception: Setup class failed

----------------------------------------------------------------------
Ran 0 tests in 0.016s

FAILED (errors=1)
```

Sample test module used:

```python
import unittest

class A(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        raise Exception('Setup class failed')

    def test_1(self):
        self.fail('test_1')

    def test_2(self):
        self.fail('test_2')
```

elyezer commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,279 @@ def make_template(options=None):
+            not options.get('url', None)):
+        raise CLIFactoryError('Please provide valid custom repo URL.')
+    # Create new organization and lifecycle environment if needed
+    if not options.get('organization-id', None):
```
lines 1584-1587 can be rewritten to

```python
org_id = options.get('organization-id', make_org()['id'])
```

elyezer commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,279 @@ def make_template(options=None):
+        org_id = make_org()['id']
+    else:
+        org_id = options.get('organization-id', None)
+    if not options.get('lifecycle-environment-id', None):
```
lines 1588-1593 can be rewritten to:

```python
env_id = options.get(
    'lifecycle-environment-id',
    make_lifecycle_environment({
        u'organization-id': org_id,
    })['id']
)
```

elyezer commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,279 @@ def make_template(options=None):
+    if result.return_code != 0:
+        raise CLIFactoryError('Failed to synchronize repository')
+    # Create CV if needed and associate repo with it
+    if not options.get('content-view-id', None):
```
```python
cv_id = options.get(
    'content-view-id', 
    make_content_view({u'organization-id': org_id})['id']
)
```

elyezer commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,279 @@ def make_template(options=None):
+    if result.return_code != 0:
+        raise CLIFactoryError('Failed to promote version to next environment')
+    # Create activation key if needed and associate content view with it
+    if not options.get('activationkey-id', None):
```
if 'activationkey-id' not in options:

elyezer commented on robottelo/cli/factory.py:
```diff
@@ -1540,3 +1545,279 @@ def make_template(options=None):
+            u'content-view-id': cv_id,
+        })
+        if result.return_code != 0:
+            raise CLIFactoryError('Failed to associate activation-key '
```
```python
raise CLIFactoryError(
    'Failed to associate activation-key with CV')
```

or if does not fit:

```python
raise CLIFactoryError(
    'Failed to associate activation-key '
    'with CV'
)
```

Right aligned code is better to read.

elyezer commented:
I have reviewed again this PR and have made some formatting related comments, they are not blockers but will be good to have.

This is the first iteration on these functions, and we can refactor them later as suggested by @sthirugn.

I'm OK on merging this PR as is and improve later.

abalakh commented:
Updated. Thanks for your feedback, guys :+1:

sthirugn commented:
Looks great. Thank you for fixing all the suggested review comments. @abalakh

sthirugn commented:
ACK

elyezer commented:
ACK
